questionpy_sdk/webserver/state.py

```diff
@@ -3,89 +3,230 @@
+
+    async def read_attempt_seed(self, question_id: str, attempt_id: str) -> int: ...
+    async def write_attempt_seed(self, question_id: str, attempt_id: str, seed: int) -> None: ...
 
-    async def read_question_state(self) -> str:
-        return await self._read_state_file(StateFilename.QUESTION_STATE)
+    async def read_attempt_score(self, question_id: str, attempt_id: str) -> ScoreModel: ...
+    async def write_attempt_score(self, question_id: str, attempt_id: str, score: ScoreModel) -> None: ...
 
-    async def write_question_state(self, question_state: str) -> None:
-        await self._write_state_file(StateFilename.QUESTION_STATE, question_state)
+    async def read_attempt_data(self, question_id: str, attempt_id: str) -> dict[str, JsonValue]: ...
+    async def write_attempt_data(self, question_id: str, attempt_id: str, data: dict[str, JsonValue]) -> None: ...
 
-    async def read_attempt_state(self) -> str:
-        return await self._read_state_file(StateFilename.ATTEMPT_STATE)
+    async def delete_attempt(self, question_id: str, attempt_id: str) -> None: ...
 
-    async def write_attempt_state(self, attempt_state: str) -> None:
-        await self._write_state_file(StateFilename.ATTEMPT_STATE, attempt_state)
 
-    async def read_attempt_seed(self) -> int:
-        return int(await self._read_state_file(StateFilename.ATTEMPT_SEED))
+class FilesystemStateManager:
+    """Handles asynchronous reading, writing, and deletion of question package state files.
 
-    async def write_attempt_seed(self, seed: int) -> None:
-        await self._write_state_file(StateFilename.ATTEMPT_SEED, str(seed))
+    Manages persistence of state within the specified directory. All file operations are non-blocking
+    and run in a thread pool.
+
+    Args:
+        package_state_dir: Directory where state files are stored.
+    """
+
+    class StateFilename(StrEnum):
+        QUESTION_STATE = "question_state.txt"
+        ATTEMPT_STATE = "attempt_state.txt"
+        ATTEMPT_SEED = "attempt_seed.txt"
+        ATTEMPT_SCORE = "score.json"
+        ATTEMPT_DATA = "attempt_data.json"
+
+    def __init__(self, package_state_dir: Path) -> None:
+        self._package_state_dir = package_state_dir
 
-    async def read_score(self) -> ScoreModel:
-        score_json = await self._read_state_file(StateFilename.SCORE)
+    async def read_question_states(self) -> dict[str, str]:
+        def _read_question_states() -> dict[str, str]:
+            try:
+                dir_iter = self._package_state_dir.iterdir()
+            except FileNotFoundError:
+                # Package dir might not have been created yet
+                return {}
+            else:
+                return {
+                    str(path.name): (path / self.StateFilename.QUESTION_STATE).read_text()
+                    for path in dir_iter
+                    if path.is_dir() and re.match(ID_RE, str(path.name))
+                }
```

Comment:
1. `iterdir()` schmeißt interessanterweise, im Gegensatz zu Python 3.13, unter Python 3.12.2 keinen `FileNotFoundError`. Ich bin gestern durch Zufall darauf gestoßen und habe in [den Docs](https://docs.python.org/3/library/pathlib.html#pathlib.Path.iterdir) auch nichts dazu gefunden, dort steht nur:
`If the path is not a directory or otherwise inaccessible, OSError is raised.`
Ohne try-except-else ist es mMn auch etwas übersichtlicher.

3.  `path.name` ist schon ein `str`

```suggestion
            if not self._package_state_dir.exists():
                return {}

            return {
                path.name: (path / self.StateFilename.QUESTION_STATE).read_text()
                for path in self._package_state_dir.iterdir()
                if path.is_dir() and re.match(ID_RE, path.name)
            }
```

Reply:
> 1. `iterdir()` schmeißt interessanterweise, im Gegensatz zu Python 3.13, unter Python 3.12.2 keinen `FileNotFoundError`.

Wie kommst du genau darauf? Beide Versionen verhalten sich gleich, was mich sonst auch eher wundern würde:


```py
# test.py
from pathlib import Path

try:
    Path("/does/not/exist").iterdir()
except OSError as e:
    print(type(e).__name__, e)
```

```sh
$ python --version
Python 3.12.2
$ python test.py
FileNotFoundError [Errno 2] No such file or directory: '/does/not/exist'
```

> Ohne try-except-else ist es mMn auch etwas übersichtlicher.

try/catch ist eigentlich die idiomatische Variante: *it is easier to ask forgiveness than permission*

Die Varianten sind übrigens auch nicht equivalent.

- Die if-Variante führt eine Race-Condition ein, die zu einer unbehandelten Exception führen kann:

```py
    if not self._package_state_dir.exists():
        return {}

    # zwischen den Calls verschwindet das Verzeichnis

    return {
        path.name: (path / self.StateFilename.QUESTION_STATE).read_text()
        for path in self._package_state_dir.iterdir()  # 💥 FileNotFoundError
        if path.is_dir() and re.match(ID_RE, path.name)
    }
```

- Die if-Variante braucht einen zusätzlichen System-Call.

Reply:
> 2\. `path.name` ist schon ein `str`

:+1:

Reply:
> Wie kommst du darauf?

Mir ist das aufgefallen, da ich den Ordner manuell gelöscht hatte und der Server dann beim Lesen der Question-State-Files einen `FileNotFoundError` geschmissen hat.
Ich habe es gerade nochmal mit deinem Beispiel, `pyenv`, sowie der aktuellsten Python 3.12 Version und der ersten Python 3.13 Version getestet:
```shell
$ pyenv shell 3.12.11

$ python --version
Python 3.12.11

$ python test.py

$ pyenv shell 3.13.0

$ python --version
Python 3.13.0

$ python test.py
FileNotFoundError [Errno 2] No such file or directory: '/does/not/exist'
```

Das scheint wohl ein Bug zu sein.

> try/catch ist eigentlich die idiomatische Variante: _it is easier to ask forgiveness than permission_

Da gebe ich dir grundsätzlich Recht.

> Die if-Variante führt eine Race-Condition ein, die zu einer unbehandelten Exception führen kann

Race-Conditions existieren auch im bestehenden Code und ich denke auch nicht, dass wir das ohne größeren Aufwand vermeiden können und sollten:

```python
            try:
                dir_iter = self._package_state_dir.iterdir()
            except FileNotFoundError:
                # Package dir might not have been created yet
                return {}

            # zwischen den Calls verschwindet das Verzeichnis

            return {
                path.name: (path / self.StateFilename.QUESTION_STATE).read_text()  # 💥 FileNotFoundError
                for path in dir_iter
                if path.is_dir() and re.match(ID_RE, path.name)
            }
```

> So sieht der Code jetzt übrigens aus, der else-Zweig war ja eigentlich nicht nötig und ist vielleicht so auch ein bisschen lesbarer/übersichtlicher?

Finde ich auch lesbarer. Ich weiß nicht wie verbreitet der `Path.iterdir()`-Bug ist, da wir Python 3.12 (EOL: `2028-10`) unterstützen, sollten wir vielleicht trotzdem eine andere Lösung wählen. Ich weiß auch nicht, ob ein potentieller Fix als Security-Fix durchgehen würde.

Ich habe auch gerade mit verschiedenen Online-Tools (z.B. [hier (Python 3.12.3)](https://www.onlinegdb.com/online_python_compiler)) und Python < 3.13 den Test-Code getestet und nie wurde ein Exception geschmissen.

Reply:
Ich habe gerade einen offenen [Bug-Report](https://github.com/python/cpython/issues/129871) gefunden.

Reply:
> Das scheint wohl ein Bug zu sein.

It's a feature! https://github.com/python/cpython/issues/78722

Wie dem auch sei, solange wir noch Python-Versionen <3.13 unterstützen, ist der pragmatische Umgang, den Iterator in eine `list()` zu wrappen, so dass er direkt ausgewertet wird.

Auf jeden Fall gut, dass dir das aufgefallen ist.

4511baa6f6f70eb4bf807ba9e775805c0886f003

> Finde ich auch lesbarer.

6dc55fa1d4ca599d049319bfd4e796daf8f2787e